lib/engine.js

```diff
@@ -393,77 +397,58 @@ class Engine {
-        else if(!downloadAsset && input.path) {
-            const preSignedUrl = await this.getPresignedUrl(input.path);
-            // TODO: remove this log
-            debug('generated presigned url', preSignedUrl);
-            input.url = preSignedUrl;
-            transformerInput = new Asset(input, input.path);
-
-        } // asset is an intermediate rendition from previous transformer and is already available in input.path
-        else if(input.path) {
-            transformerInput = new Asset(input, input.path);
+        if (input && !input.url && !input.path) {
+            throw new GenericError('No source file accessible.', 'prepareTransformer');
         }
 
-        if (!transformerInput || !transformerInput.path) {
-            throw new GenericError('No source file accessible.');
+        let protocol = '';
+        if (input.url) {
+            if (!validDataUrl(input.url) && !validUrl.isHttpsUri(input.url)) {
+                throw new SourceUnsupportedError(`${input.url} must be a valid https url or datauri`);
+            }
+            protocol = new URL(input.url).protocol;
+        }
+    
+        debug(`sourceType: ${input.sourceType}`);
+        if (input.sourceType === TransformerSourceType.URL) {
+            if (input.url && (protocol === 'https:')) {
+                debug(`asset url available : dont need to do anything`);
+                transformerInput = new Asset(input);
+            } else if (input.url && protocol === 'data:') {
+                debug(`asset url available as data URI : upload asset to a temp cloud storage and generate a presigned URL`);
+                transformerInput = await getSource(input, transformerDirectory.base, true);
+            } else {
+                debug(`asset url not available : generated presigned url`);
+                try {
+                    const preSignedUrl = await datauri.getPreSignedUrl(input.path);
+                    input.url = preSignedUrl;
+                    transformerInput = new Asset(input, input.path);
+                } catch (error) {
+                    console.log(error);
+                    throw new GenericError(`Fail generating presigned url.`, 'prepareTransformer');
+                }
+            }
+        } else { // defaults to assuming the asset is needed locally, treated as sourceType === 'LOCAL'
+            if (input.path) {
+                debug(`asset path available : dont need to do anything`);
+                transformerInput = new Asset(input, input.path);
+            } else {
+                debug(`asset path not available : download asset`);
+                // TODO: metric here is not really accurate when input.url is datauri, 
+                //       because it also includes upload duration
+                this[INTERNAL].timers.download.start();
+                // download input to <baseDirectory>/<stepNumber>-<transformerName>/inputfile and set input.path
```

Comment:
is it possible to have `datauri` in this flow ?

Reply:
yes, but getSource takes care of it